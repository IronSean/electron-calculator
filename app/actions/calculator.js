// @flow
import type { GetState, Dispatch } from '../reducers/types';

export const ADD_TO_TOTAL = 'ADD_TO_TOTAL';
export const MULTIPLY_TOTAL = 'MULTIPLY_TOTAL';
export const CLEAR_TOTAL = 'CLEAR_TOTAL';

export function addToTotal(value: number) {
  return {
    type: ADD_TO_TOTAL,
    payload: value
  };
}

export function multiplyTotal(value: number) {
  return {
    type: MULTIPLY_TOTAL,
    payload: value
  };
}

export function subtractFomTotal(value: number) {
  return (dispatch: Dispatch) => {
    dispatch(addToTotal(-value));
  };
}

export function divideTotal(value: number) {
  return (dispatch: Dispatch) => {
    if (value === 0) {
      return;
    }

    dispatch(multiplyTotal(1 / value));
  };
}

export function clearTotal() {
  return {
    type: CLEAR_TOTAL
  };
}
