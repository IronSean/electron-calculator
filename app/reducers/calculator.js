// @flow
import {
  ADD_TO_TOTAL,
  MULTIPLY_TOTAL,
  CLEAR_TOTAL
} from '../actions/calculator';
import type { Action } from './types';

export default function calculator(state: number = 0, action: Action) {
  switch (action.type) {
    case ADD_TO_TOTAL:
      return state + action.payload;
    case MULTIPLY_TOTAL:
      return state * action.payload;
    case CLEAR_TOTAL:
      return 0;
    default:
      return state;
  }
}
