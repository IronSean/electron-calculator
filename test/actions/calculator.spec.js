import { spy } from 'sinon';
import * as actions from '../../app/actions/calculator';

describe('actions', () => {
  it('should addToTotal should create ADD_TO_TOTAL action with value as payload', () => {
    expect(actions.addToTotal(12)).toMatchSnapshot();
  });

  it('should multiplyTotal should create MULTIPLY_TOTAL action with value as payload', () => {
    expect(actions.multiplyTotal(12)).toMatchSnapshot();
  });

  it('should subtractTotal should create ADD_TO_TOTAL action with negative value as payload', () => {
    const value = 12;
    const fn = actions.subtractFomTotal(value);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(
      dispatch.calledWith({ type: actions.ADD_TO_TOTAL, payload: -value })
    ).toBe(true);
  });

  it('should divideTotal should create MULTIPLY_TOTAL action with inverse value as payload', () => {
    const value = 12;
    const fn = actions.divideTotal(value);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(
      dispatch.calledWith({ type: actions.MULTIPLY_TOTAL, payload: 1 / value })
    ).toBe(true);
  });

  it('should divideTotal shouldnt create MULTIPLY_TOTAL action if value is zero', () => {
    const fn = actions.divideTotal(0);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.called).toBe(false);
  });

  // There's no nice way to test this at the moment...
  it('should clearTotal should create CLEAR_TOTAL action', () => {
    expect(actions.clearTotal()).toMatchSnapshot();
  });
});
