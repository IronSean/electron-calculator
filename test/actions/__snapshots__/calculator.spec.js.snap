// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should addToTotal should create ADD_TO_TOTAL action with value as payload 1`] = `
Object {
  "payload": 12,
  "type": "ADD_TO_TOTAL",
}
`;

exports[`actions should clearTotal should create CLEAR_TOTAL action 1`] = `
Object {
  "type": "CLEAR_TOTAL",
}
`;

exports[`actions should multiplyTotal should create MULTIPLY_TOTAL action with value as payload 1`] = `
Object {
  "payload": 12,
  "type": "MULTIPLY_TOTAL",
}
`;
